<?xml version="1.0"?>
<robot name="racecar" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:include filename="$(find racecar_description)/urdf/macros.xacro" />

  <!-- Add chassis and it's inertia link -->

  <link name="base_link" />

  <link name="chassis">
    <visual>
      <xacro:chassis_geometry/>
      <material name="blue" />
    </visual>
  </link>

  <joint name="base_link_joint" type="fixed">
    <origin xyz="0 0 0.05" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="chassis" />
  </joint>

  <link name="chassis_inertia">
    <inertial>
      <xacro:chassis_inertial_params />
    </inertial>
  </link>

  <joint name="chassis_inertia_joint" type="fixed">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="chassis" />
    <child link="chassis_inertia" />
  </joint>

  <!-- Add the left rear wheel with its joints and tranmissions -->

  <link name="left_rear_wheel">
    <inertial>
      <xacro:left_wheels_inertial_params />
    </inertial>
    <visual>
      <xacro:left_rear_wheel_geometry />
      <material name="black" />
    </visual>
    <collision>robot
      <xacro:left_wheels_collision_geometry />
    </collision>
  </link>

  <joint name="left_rear_wheel_joint" type="continuous">
    <origin xyz="0 0.1 0" rpy="1.5708 0 0" />
    <parent link="chassis" />
    <child link="left_rear_wheel" />
    <axis xyz="0 0 -1" />
    <limit effort="10" velocity="100" />
  </joint>

  <xacro:wheel_transmission name="left_rear_wheel" />

  <!-- Add the right rear wheel with its joints and tranmissions -->

  <link name="right_rear_wheel">
    <inertial>
      <xacro:right_wheels_inertial_params />
    </inertial>
    <visual>
      <xacro:right_rear_wheel_geometry />
      <material name="black" />
    </visual>
    <collision>
      <xacro:right_wheels_collision_geometry />
    </collision>
  </link>

  <joint name="right_rear_wheel_joint" type="continuous">
    <origin xyz="0 -0.1 0" rpy="1.5708 0 0" />
    <parent link="chassis" />
    <child link="right_rear_wheel" />
    <axis xyz="0 0 -1" />
    <limit effort="10" velocity="100" />
  </joint>

  <xacro:wheel_transmission name="right_rear_wheel" />

  <!-- Add the left steering hinge with its joints and tranmissions -->

  <link name="left_steering_hinge">
    <inertial>
      <xacro:steering_hinge_inertial_params />
    </inertial>
    <visual>
      <xacro:left_steering_hinge_geometry />
      <material name="grey" />
    </visual>
  </link>

  <joint name="left_steering_hinge_joint" type="revolute">
    <origin xyz="0.325 0.1 0" rpy="0 1.5708 0" />
    <parent link="chassis" />
    <child link="left_steering_hinge" />
    <axis xyz="-1 0 0" />
    <limit lower="-1.0" upper="1.0" effort="10" velocity="100" />
  </joint>

  <xacro:steering_hinge_transmission name="left_steering_hinge" />

  <!-- Add the right steering hinge with its joints and tranmissions -->

  <link name="right_steering_hinge">
    <inertial>
      <xacro:steering_hinge_inertial_params />
    </inertial>
    <visual>
      <xacro:right_steering_hinge_geometry />
      <material name="grey" />
    </visual>
  </link>

  <joint name="right_steering_hinge_joint" type="continuous">
    <origin xyz="0.325 -0.1 0" rpy="0 1.5708 0" />
    <parent link="chassis" />
    <child link="right_steering_hinge" />
    <axis xyz="-1 0 0" />
    <limit lower="-1.0" upper="1.0" effort="10" velocity="100" />
  </joint>

  <xacro:steering_hinge_transmission name="right_steering_hinge" />

  <!-- Add the left front wheel with its joints and tranmissions -->

  <link name="left_front_wheel">
    <inertial>
      <xacro:left_wheels_inertial_params />
    </inertial>
    <visual>
      <xacro:left_front_wheel_geometry />
      <material name="black" />
    </visual>
    <collision>
      <xacro:left_wheels_collision_geometry />
    </collision>
  </link>

  <joint name="left_front_wheel_joint" type="continuous">
    <origin xyz="0 0 0" rpy="1.5708 0 0" />
    <parent link="left_steering_hinge" />
    <child link="left_front_wheel" />
    <axis xyz="0 0 -1" />
    <limit effort="10" velocity="100" />
  </joint>

  <xacro:wheel_transmission name="left_front_wheel" />

  <!-- Add the left front wheel with its joints and tranmissions -->

  <link name="right_front_wheel">
    <inertial>
      <xacro:right_wheels_inertial_params />
    </inertial>
    <visual>
      <xacro:right_front_wheel_geometry />
      <material name="black" />
    </visual>
    <collision>
      <xacro:right_wheels_collision_geometry />
    </collision>
  </link>

  <joint name="right_front_wheel_joint" type="continuous">
    <origin xyz="0 0 0" rpy="1.5708 0 0" />
    <parent link="right_steering_hinge" />
    <child link="right_front_wheel" />
    <axis xyz="0 0 -1" />
    <limit effort="10" velocity="100"/>
  </joint>
  
  <xacro:wheel_transmission name="right_front_wheel" />

  <!-- Add Hokuyo laser scanner -->

  <link name="laser">
    <inertial>
      <xacro:hokuyo_inertial_params />
    </inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://racecar_description/meshes/hokuyo.dae"/>
        <material name="grey" />
      </geometry>
    </visual>
  </link>

  <joint name="hokuyo_joint" type="fixed">
    <origin xyz="0.145 0.0 0.158" rpy="0 0 0"/>
    <parent link="chassis" />
    <child link="laser" />
    <axis xyz="0 0 1" />
  </joint>
  
  <!-- IMU config -->
  
  <link name="base_imu_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.03 0.04 0.006"/>
      </geometry>
      <material name="red"/>
    </visual>
  </link>
  
   <joint name="base_imu_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0.255 0 0.078" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="base_imu_link"/>
  </joint>

 <!-- Basler_camera config -->

  <link name="base_basler_camera_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.0293 0.029 0.029"/>
      </geometry>
      <material name="gray"/>
    </visual>
  </link>

   <joint name="base_basler_camera_joint" type="fixed">
    <axis xyz="0 0 0" />
    <origin xyz="0.145 0.0 0.208" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="base_basler_camera_link"/>
  </joint>


<!-- full zed camera config -->

  <joint name="zed_camera_joint" type="fixed">
    <origin xyz="0.295 0 0.138" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="zed_center"/>
    <axis xyz="0 0 1" />
  </joint>

	
  
   <link name="zed_left_camera">
        <visual>
            <origin xyz="0 0 0" rpy="0 1.57 0"/>
            <geometry>
                <cylinder radius="0.007" length=".031"/>
            </geometry>
            <material name="dark_grey">
                <color rgba="0.1 0.1 0.1 1"/>
            </material>
        </visual>
    </link>
    <link name="zed_center">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://racecar_description/meshes/ZED.stl" />
            </geometry>
            <material name="light_grey">
                <color rgba="0.8 0.8 0.8 0.8"/>
            </material>
        </visual>
    </link>
    <link name="zed_right_camera">
        <visual>
            <origin xyz="0 0 0" rpy="0 1.57 0"/>
            <geometry>
                <cylinder radius="0.007" length=".031"/>
            </geometry>
            <material name="dark_grey">
                <color rgba="0.1 0.1 0.1 1"/>
            </material>
        </visual>
    </link>
    <link name="zed_depth_camera" />

    <joint name="zed_left_camera_joint" type="fixed">
        <parent link="zed_center"/>
        <child link="zed_left_camera"/>
        <origin xyz="0 0.06 0" rpy="0 0 0" />
    </joint>
    <joint name="zed_depth_camera_joint" type="fixed">
        <parent link="zed_left_camera"/>
        <child link="zed_depth_camera"/>
        <origin xyz="0 0 0" rpy="-1.5707963267948966 0 -1.5707963267948966" />
    </joint>

    <joint name="zed_right_camera_joint" type="fixed">
        <parent link="zed_center"/>
        <child link="zed_right_camera"/>
        <origin xyz="0 -0.06 0" rpy="0 0 0" />
    </joint>

  <!-- Add the remaining xacros -->
  <xacro:include filename="$(find racecar_description)/urdf/racecar.gazebo" />
  <xacro:include filename="$(find racecar_description)/urdf/materials.xacro" />


</robot>
